### ISSUE TITLE
Decorator/Attribute Semantic Token Type

### ISSUE THREAD

**Issue by justinytchen** [microsoft/language-server-protocol#1067]
Is there a particular reason why there's currently not a semantic token type for decorators or attributes? If not, would it be possible to create a one?

**Comment by TylerLeonhardt** [comment]
It doesn't seem like TypeScript has one applied today:
![image](https://user-images.githubusercontent.com/2644648/89953653-5c8d1e00-dbe4-11ea-80ca-308605c25f24.png)

It'd be great to see a semantic token type for this.

For PowerShell it would be very useful to differentiate between Types and Attributes:
![image](https://user-images.githubusercontent.com/2644648/89954271-a32f4800-dbe5-11ea-8fe1-bf01c25bdd59.png)

**Comment by Eskibear** [comment]
In Java, there is also annotation type.
e.g.
```java
@SuppressWarnings("restriction")
public class Foo {
// ...
}
```

Just FYI, Java language server defines a new `annotation` type for the case. And VSCode Java extension also declares [custom token types](https://github.com/redhat-developer/vscode-java/blob/715bf2667352060adf79d98190a7bb2c23f7b3b6/package.json#L37-L47) , e.g. `annotation` inherits from standard token type `type`, to leverage highlighting from most themes.

**Comment by TylerLeonhardt** [comment]
@Eskibear thanks! I'd still love to see a specific `SemanticTokenType so that the built in themes to VS Code could include a specific color for Decorator/Attriubute/Annotation.

**Comment by rcjsuen** [comment]
> @Eskibear thanks! I'd still love to see a specific `SemanticTokenType so that the built in themes to VS Code could include a specific color for Decorator/Attriubute/Annotation.

@TylerLeonhardt Just to be clear, the protocol declaring it has no bearing on what happens in VS Code. Code changes would still be required on that end to support this.

**Comment by TylerLeonhardt** [comment]
@rcjsuen yes but it's more likely that VS Code would be willing to include coloring in their default themes for a semantic token for Decorator/Attriubute/Annotation defined by LSP.

They would be less inclined to colorize a PowerShell/.NET specific custom semantic token type for "Attribute".

Since a lot of languages have this concept (Decorator/Attriubute/Annotation) it would make sense to have it included in the LSP so that editors could generically support colorizing it.

**Comment by rcjsuen** [comment]
> @rcjsuen yes but it's more likely that VS Code would be willing to include coloring in their default themes for a semantic token for Decorator/Attriubute/Annotation defined by LSP.

Looks like that has already been raised now that I've done some digging (https://github.com/microsoft/vscode/issues/97063#issuecomment-673947951).

**Comment by dbaeumer** [comment]
One thing to note is that the list of tokens is not fixed in the protocol. If a client announces an additional token a server can still use it even if it is not defined in the LSP protocol.

**Comment by puremourning** [comment]
> One thing to note is that the list of tokens is not fixed in the protocol. If a client announces an additional token a server can still use it even if it is not defined in the LSP protocol.

But this mechanism strictly couples clients and servers and makes it essentially impossible to write a generic client implementation.

**Comment by TylerLeonhardt** [comment]
@dbaeumer please see my response to that a couple comments up: https://github.com/microsoft/language-server-protocol/issues/1067#issuecomment-673115647

**Comment by dbaeumer** [comment]
I will coordinate this with VS Code.

**Comment by asklar** [comment]
FWIW MIDL (interface definition language) would like to use this. microsoft/vscode-languageserver-node#754
